File: benchmarks/sqlite_bench/sqlite_bench.js
// SQLite benchmark - JavaScript (Node.js with better-sqlite3)
// Tests: create table, insert, query, update, delete

const Database = require('better-sqlite3');

const n = parseInt(process.argv[2]) || 1000;

// Use in-memory database for benchmarking
const db = new Database(':memory:');

// Create table
db.exec('CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT, email TEXT, score INTEGER)');

// Insert N rows using transaction for performance
const insert = db.prepare('INSERT INTO users (name, email, score) VALUES (?, ?, ?)');
const insertMany = db.transaction((items) => {
    for (const item of items) insert.run(item.name, item.email, item.score);
});

const items = [];
for (let i = 0; i < n; i++) {
    items.push({ name: `user${i}`, email: `user${i}@example.com`, score: i * 10 });
}
insertMany(items);

// Query all rows
const rows = db.prepare('SELECT * FROM users').all();
let totalScore = 0;
for (const row of rows) {
    totalScore += row.score;
}

// Update all rows
const update = db.prepare('UPDATE users SET score = score + 1 WHERE id = ?');
const updateMany = db.transaction(() => {
    for (let i = 0; i < n; i++) {
        update.run(i + 1);
    }
});
updateMany();

// Delete all rows
db.exec('DELETE FROM users');

db.close();

console.log(totalScore);

File: benchmarks/sqlite_bench/sqlite_bench.c
// SQLite benchmark - C
// Tests: create table, insert, query, update, delete

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sqlite3.h>

int main(int argc, char *argv[]) {
    int n = 1000;
    if (argc > 1) {
        n = atoi(argv[1]);
    }

    sqlite3 *db;
    sqlite3_stmt *stmt;
    char sql[256];
    int rc;
    long long total_score = 0;

    // Use in-memory database for benchmarking
    rc = sqlite3_open(":memory:", &db);
    if (rc != SQLITE_OK) {
        fprintf(stderr, "Cannot open database: %s\n", sqlite3_errmsg(db));
        return 1;
    }

    // Create table
    rc = sqlite3_exec(db, "CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT, email TEXT, score INTEGER)", NULL, NULL, NULL);
    if (rc != SQLITE_OK) {
        fprintf(stderr, "SQL error: %s\n", sqlite3_errmsg(db));
        sqlite3_close(db);
        return 1;
    }

    // Begin transaction for inserts
    sqlite3_exec(db, "BEGIN TRANSACTION", NULL, NULL, NULL);

    // Prepare insert statement
    rc = sqlite3_prepare_v2(db, "INSERT INTO users (name, email, score) VALUES (?, ?, ?)", -1, &stmt, NULL);
    if (rc != SQLITE_OK) {
        fprintf(stderr, "Prepare error: %s\n", sqlite3_errmsg(db));
        sqlite3_close(db);
        return 1;
    }

    // Insert N rows
    char name[32], email[64];
    for (int i = 0; i < n; i++) {
        snprintf(name, sizeof(name), "user%d", i);
        snprintf(email, sizeof(email), "user%d@example.com", i);

        sqlite3_bind_text(stmt, 1, name, -1, SQLITE_TRANSIENT);
        sqlite3_bind_text(stmt, 2, email, -1, SQLITE_TRANSIENT);
        sqlite3_bind_int(stmt, 3, i * 10);

        sqlite3_step(stmt);
        sqlite3_reset(stmt);
    }
    sqlite3_finalize(stmt);

    // Commit transaction
    sqlite3_exec(db, "COMMIT", NULL, NULL, NULL);

    // Query all rows
    rc = sqlite3_prepare_v2(db, "SELECT * FROM users", -1, &stmt, NULL);
    if (rc != SQLITE_OK) {
        fprintf(stderr, "Query prepare error: %s\n", sqlite3_errmsg(db));
        sqlite3_close(db);
        return 1;
    }

    while (sqlite3_step(stmt) == SQLITE_ROW) {
        total_score += sqlite3_column_int(stmt, 3);
    }
    sqlite3_finalize(stmt);

    // Begin transaction for updates
    sqlite3_exec(db, "BEGIN TRANSACTION", NULL, NULL, NULL);

    // Prepare update statement
    rc = sqlite3_prepare_v2(db, "UPDATE users SET score = score + 1 WHERE id = ?", -1, &stmt, NULL);
    if (rc != SQLITE_OK) {
        fprintf(stderr, "Update prepare error: %s\n", sqlite3_errmsg(db));
        sqlite3_close(db);
        return 1;
    }

    // Update all rows
    for (int i = 0; i < n; i++) {
        sqlite3_bind_int(stmt, 1, i + 1);
        sqlite3_step(stmt);
        sqlite3_reset(stmt);
    }
    sqlite3_finalize(stmt);

    // Commit transaction
    sqlite3_exec(db, "COMMIT", NULL, NULL, NULL);

    // Delete all rows
    sqlite3_exec(db, "DELETE FROM users", NULL, NULL, NULL);

    sqlite3_close(db);

    printf("%lld\n", total_score);
    return 0;
}
